File: frontend/lib/schemas/style.ts
import { z } from "zod";

/**
 * Zod 版 StyleSchema。單一事實來源是 backend Pydantic。
 * Sprint 3 會接 `json-schema-to-zod` 自動 codegen 取代手寫。
 */

const hexColor = z.string().regex(/^#[0-9A-Fa-f]{6}$/);

export const materialSchema = z.object({
  type: z.string(),
  finish: z.enum(["matte", "satin", "gloss"]).default("matte"),
  tone: z.enum(["light", "medium", "dark"]).default("medium"),
  color: hexColor.nullable().optional(),
});

export const lightingSchema = z.object({
  sun_kelvin: z.number().int().min(2500).max(10000),
  direction: z.enum(["N", "NE", "E", "SE", "S", "SW", "W", "NW"]),
  intensity: z.number().min(0).max(3),
  ambient: z.enum(["soft", "hard", "overcast", "golden-hour"]),
});

export const cameraSchema = z.object({
  type: z.enum(["eye-level", "axonometric", "aerial", "custom"]).default("eye-level"),
  fov: z.number().int().default(50),
  height_mm: z.number().int().default(1600),
});

export const styleSchema = z.object({
  color_palette: z.array(hexColor).min(3).max(5),
  floor: materialSchema,
  wall: materialSchema,
  ceiling: materialSchema,
  lighting: lightingSchema,
  furniture_language: z.enum([
    "nordic-minimal",
    "wabi-sabi",
    "industrial",
    "modern",
    "luxury",
    "american-country",
    "french",
  ]),
  camera: cameraSchema,
});

export type StyleSchema = z.infer<typeof styleSchema>;

File: backend/app/schemas/style.py
"""StyleSchema — 單一事實來源（Pydantic → JSON Schema → Zod）。

規格 §4.3。使用 Pydantic v2 的 `Annotated[str, StringConstraints]`
取代 v1 的 `constr`。
"""

from typing import Annotated, Literal

from pydantic import BaseModel, Field, StringConstraints

HexColor = Annotated[str, StringConstraints(pattern=r"^#[0-9A-Fa-f]{6}$")]

FurnitureLanguage = Literal[
    "nordic-minimal",
    "wabi-sabi",
    "industrial",
    "modern",
    "luxury",
    "american-country",
    "french",
]
Direction = Literal["N", "NE", "E", "SE", "S", "SW", "W", "NW"]
Ambient = Literal["soft", "hard", "overcast", "golden-hour"]
CameraType = Literal["eye-level", "axonometric", "aerial", "custom"]
Finish = Literal["matte", "satin", "gloss"]
Tone = Literal["light", "medium", "dark"]


class Material(BaseModel):
    type: str = Field(..., description="oak, latex_paint, marble ...")
    finish: Finish = "matte"
    tone: Tone = "medium"
    color: HexColor | None = None


class Lighting(BaseModel):
    sun_kelvin: int = Field(..., ge=2500, le=10000)
    direction: Direction
    intensity: float = Field(..., ge=0, le=3)
    ambient: Ambient


class Camera(BaseModel):
    type: CameraType = "eye-level"
    fov: int = 50
    height_mm: int = 1600


class StyleSchema(BaseModel):
    color_palette: list[HexColor] = Field(..., min_length=3, max_length=5)
    floor: Material
    wall: Material
    ceiling: Material
    lighting: Lighting
    furniture_language: FurnitureLanguage
    camera: Camera
